openpype/lib/abstract_load_template.py

```diff
@@ -0,0 +1,305 @@
+
+        representations_by_id = {
+            representation['_id']: representation
+            for asset in version_repres.values()
+            for subset in asset['subsets'].values()
+            for representation in subset['version']['repres']}
+
+        context_representations_by_id = dict()
+        linked_representations_by_id = dict()
+        for k in representations_by_id.keys():
+            asset = representations_by_id[k]['context']['asset']
+            if asset == current_asset:
+                context_representations_by_id[k] = representations_by_id[k]
+            else:
+                linked_representations_by_id[k] = representations_by_id[k]
+
+        placeholders = map(placeholder_class, self.get_template_nodes())
+        valid_placeholders = filter(placeholder_class.is_valid, placeholders)
+        sorted_placeholders = sorted(valid_placeholders,
+                                     key=placeholder_class.order)
+
+        loaded_assets = set()
+        for placeholder in sorted_placeholders:
+            if placeholder.data['builder_type'] == 'context_asset':
+                representations_by_id = context_representations_by_id
+                assets_to_load.add(current_asset)
+            else:
+                representations_by_id = linked_representations_by_id
+            for items in representations_by_id.items():
+                representation_id, representation = items
+                if not placeholder.is_repres_valid(representation):
+                    continue
+                container = avalon.api.load(
+                    loaders_by_name[placeholder.loader],
+                    representation_id)
+                placeholder.parent_in_hierarchy(container)
+                loaded_assets.add(representation['context']['asset'])
+            placeholder.clean()
+
+        if not loaded_assets == assets_to_load:
+            unloaded = assets_to_load - loaded_assets
+            print("Error found while loading {}".format(unloaded))
+            print("It's possible that a needed asset wasn't published")
+            print("or that the build template is malformed, "
+                  "continue at your own risks.")
+
+    def import_template(self, template_path):
+        """
+        Import template in current dcc
+
+        Args:
+            template_path (str): fullpath to current task and
+                dcc's template file
+
+        Return:
+            None
+        """
+        raise NotImplementedError
+
+    def get_template_nodes(self):
+        """
+        Returning a list of nodes acting as DCC placeholders for
+        templating. The data representation is by user.
+        AbstractLoadTemplate (and LoadTemplate) won't directly manipulate nodes
+
+        Args :
+            None
+
+        Returns:
+            list(AnyNode): Solved template path
+        """
+        raise NotImplementedError
+
+
+class AbstractPlaceholder:
```

Comment:
If it's abstract then should use `ABCMeta` as metaclass and have `@abstractmethod` decorators on abstract methods.